cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperatorTest.java

```diff
@@ -90,65 +99,661 @@
+                kr).setHandler(context.succeeding(v -> context.verify(() -> {
+                    // the resource moved from New to PendingProposal (due to the configured Mock server pending calls)
+                    KafkaRebalance kr1 = Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).withName(RESOURCE_NAME).get();
+                    assertState(kr1, KafkaRebalanceAssemblyOperator.State.PendingProposal);
+                }))).compose(v -> {
+                    // trigger another reconcile to process the PendingProposal state
+                    KafkaRebalance kr1 = Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).withName(RESOURCE_NAME).get();
+
+                    return kcrao.reconcileRebalance(
+                            new Reconciliation("test-trigger", KafkaRebalance.RESOURCE_KIND, CLUSTER_NAMESPACE, RESOURCE_NAME),
+                            kr1);
+                }).setHandler(context.succeeding(v -> context.verify(() -> {
+                    // the resource moved from New to ProposalReady
+                    KafkaRebalance kr1 = Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).withName(RESOURCE_NAME).get();
+                    assertState(kr1, KafkaRebalanceAssemblyOperator.State.ProposalReady);
+                    context.completeNow();
+                })));
+    }
+
+    @Test
+    public void testNewToPendingProposalToStoppedRebalance(Vertx vertx, VertxTestContext context) throws IOException, URISyntaxException {
+
+        // Setup the rebalance endpoint with the number of pending calls before a response is received.
+        MockCruiseControl.setupCCRebalanceResponse(ccServer, 5);
+
+        KafkaRebalance kr =
+                createKafkaRebalance(CLUSTER_NAMESPACE, CLUSTER_NAME, RESOURCE_NAME, new KafkaRebalanceSpecBuilder().build());
+
+        Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).create(kr);
+
+        when(mockKafkaOps.getAsync(CLUSTER_NAMESPACE, CLUSTER_NAME)).thenReturn(Future.succeededFuture(kafka));
+        mockRebalanceOperator(mockRebalanceOps, CLUSTER_NAMESPACE, RESOURCE_NAME, kubernetesClient);
+
+        kcrao.reconcileRebalance(
+                new Reconciliation("test-trigger", KafkaRebalance.RESOURCE_KIND, CLUSTER_NAMESPACE, RESOURCE_NAME),
+                kr).setHandler(context.succeeding(v -> context.verify(() -> {
+                    // the resource moved from New to PendingProposal (due to the configured Mock server pending calls)
+                    KafkaRebalance kr1 = Crds.kafkaRebalanceOperation(kubernetesClient).inNamespace(CLUSTER_NAMESPACE).withName(RESOURCE_NAME).get();
+                    assertState(kr1, KafkaRebalanceAssemblyOperator.State.PendingProposal);
+                }))).compose(v -> {
+
+                    vertx.setTimer(10000, t -> {
```

Comment:
Do this 10 seconds make the test run slow? We should avoid that if possible, and ideally avoid hard coded timeouts in tests altogether.

Reply:
We have a hardcoded polling in the Kafka Rebalance operator of 5 seconds. I am just providing the time of a couple of poll requests before annotating the resource with "stop" for testing the Stopped state.

Reply:
Yes, I understand what you're trying to do. But presumably that means it takes at least 10s for this test to run. That's fine for you right here and now, when you're focussed on these tests. But if we did that everywhere, in all the unit tests, then it becomes a problem. People stop running the unit tests because they're too slow. So the question I'm really trying to ask is: "Can you write this test so that it doesn't take 10s?" and "Can you make sure all the tests that you write don't have arbitrary sleeps in them". I know that usually makes the tests harder to write. It often makes them more robust, because you end up waiting for an observation, rather than an arbitrary time that's based on some other time configured somewhere else. With the sleep, if the reconciliation doesn't run (e.g. due to resource contraints on some CI) then the test fails. OTOH, if you wait for an observation the test is more robust because you maybe have a two minute timeout for the CI to get its act together. It also means the test runs as fast as it can.

Reply:
In this case, we have this "internal" polling in the rebalance operator; tbh I don't see a way to "observe" something. If the operator was relying on the reconcile loop then I had control on it and applying the annotation before doing a reconcile. Right now I can avoid the timer here and applying the annotation straight in the handler but anyway the operator will get this change after 5 seconds (the hard coded one in the polling). So I think that we cannot have less then 5 seconds duration for this kind of test.

Reply:
Can't you wait to observe the internal timers interaction with the CC mock?